llvm/lib/Transforms/Yk/BlockDisambiguate.cpp

```diff
@@ -213,11 +280,37 @@ class YkBlockDisambiguate : public ModulePass {
              SuccIdx++) {
           BasicBlock *SuccBB = SI->getSuccessor(SuccIdx);
           if (SuccBB == &BB) {
-            BasicBlock *DBB = makeDisambiguationBB(Context, &BB, NewBBs);
+            BasicBlock *DBB = makeBranchDisambiguationBB(Context, &BB, NewBBs);
             SI->setSuccessor(SuccIdx, DBB);
             BB.replacePhiUsesWith(&BB, DBB);
           }
         }
+      } else if (isa<ReturnInst>(TI)) {
+        // Apply return from self-recursion disambiguation.
+        //
+        // YKFIXME: We do this even if the function not self-recursive.
```

Comment:
s/not/is not/

I also don't know what "self-recursive" means.

Reply:
A function which calls itself. I almost called it "mutual recursion", but after looking it up, that's not right. Is there a better term?

Reply:
"Directly recursive".

I suspect we also want to test "indirect recursion" (i.e. where f calls g which calls f).

Reply:
It won't tickle the particular bug we are dealing with here, but it would be good to add a test anyway.

Reply:
Works for me!

Reply:
Comments fixed in [b7346a8](https://github.com/ykjit/ykllvm/pull/27/commits/b7346a880b926338b2c159ebb0f1afc9c2fce0e1)